cooked-validators/src/Cooked/MockChain/BlockChain.hs

```diff
@@ -13,49 +15,95 @@
 
 module Cooked.MockChain.BlockChain where
 
+import qualified Cardano.Node.Emulator as Emulator
+import Control.Arrow
+import Control.Monad.Except
 import Control.Monad.Reader
 import Control.Monad.State
+import Control.Monad.Trans.Control
 import Control.Monad.Trans.Writer
+import Cooked.MockChain.GenerateTx (GenerateTxError)
 import Cooked.Output
 import Cooked.Skeleton
 import Data.Kind
 import Data.Maybe
-import qualified Ledger as Pl
-import qualified Ledger.Tx.CardanoAPI as Pl
+import qualified Ledger.Index as Ledger
+import qualified Ledger.Slot as Ledger
+import qualified Ledger.Tx as Ledger
+import qualified Ledger.Tx.CardanoAPI as Ledger
 import ListT
 import Optics.Core
+import qualified Plutus.Script.Utils.Scripts as Pl
 import qualified Plutus.V2.Ledger.Api as PV2
-import qualified PlutusTx as Pl
+import qualified Plutus.V2.Ledger.Api as Pl
 import Prettyprinter (Doc)
 
 -- * BlockChain Monad
 
-class (MonadFail m) => MonadBlockChainWithoutValidation m where
-  -- | Returns a list of all currently known outputs
-  allUtxos :: m [(Pl.TxOutRef, PV2.TxOut)]
+-- | The errors that can be produced by the 'MockChainT' monad
+data MockChainError where
+  MCEValidationError :: Ledger.ValidationErrorInPhase -> MockChainError
+  MCEUnbalanceable :: String -> BalanceStage -> TxSkel -> MockChainError
+  MCENoSuitableCollateral :: MockChainError
+  MCEGenerationError :: GenerateTxError -> MockChainError
+  MCECalcFee :: MockChainError -> MockChainError
+  MCEUnknownOutRefError :: String -> PV2.TxOutRef -> MockChainError
+  MCEUnknownValidator :: String -> PV2.ValidatorHash -> MockChainError
+  MCEUnknownDatum :: String -> PV2.DatumHash -> MockChainError
+  FailWith :: String -> MockChainError
+  OtherMockChainError :: (Show err, Eq err) => err -> MockChainError
```

Comment:
Just in case you prefer the classical style instead of GADT syntax, you can still have the existential `err` as
```haskell
data MockChainError
  = ...
  | FailWith String
  | forall err. (Show err, Eq err) => OtherMockChainError err
```